### PR TITLE
5811 merge block tags

### PR DESCRIPTION
**Pull request by jdevcs**
## Description

https://github.com/web3/web3.js/issues/5811

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist for 1.x:

- [ ] I have selected the correct base branch.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [ ] Any dependent changes have been merged and published in downstream modules.
- [ ] I ran `npm run dtslint` with success and extended the tests and types if necessary.
- [ ] I ran `npm run test:cov` and my test cases cover all the lines and branches of the added code.
- [ ] I ran `npm run build` with success.
- [ ] I have tested the built `dist/web3.min.js` in a browser.
- [ ] I have tested my code on the live network.
- [ ] I have checked the Deploy Preview and it looks correct.
- [ ] I have updated the `CHANGELOG.md` file in the root folder.

## Checklist for 4.x:

- [ ] I have selected the correct 4.x base branch.
- [ ] Within the description, the feature or issue is discussed in detail or an issue is linked.   
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have added any required tests for the changes I made 
- [ ] I have made corresponding changes to the documentation
- [ ] Any dependent changes have been merged and published in downstream modules.
- [ ] I ran `yarn` successfully
- [ ] I ran `yarn lint` successfully
- [ ] I ran `yarn build:web` successfully
- [ ] I ran `yarn test:unit` successfully
- [ ] I ran `yarn test:integration` successfully
- [ ] I ran `compile:contracts` successfully
- [ ] I have tested my code.
- [ ] I have updated the corresponding `CHANGELOG.md` file in the packages I have edited.

### CODE REVIEW

**Review by Aheyachy — APPROVED** [review]
[A spam message has been deleted]

File: packages/web3-types/src/eth_types.ts [hunk 1]
```diff
@@ -45,8 +45,11 @@ export enum BlockTags {
 	EARLIEST = 'earliest',
 	LATEST = 'latest',
 	PENDING = 'pending',
+	SAFE = `safe`,
+	FINALIZED = `finalized`,
```
Comment (luu-alex) [inline packages/web3-types/src/eth_types.ts]:
```suggestion
	SAFE = 'safe',
	FINALIZED = 'finalized',
```
should stick to one format

**Review by luu-alex — APPROVED** [review]

File: packages/web3-eth/CHANGELOG.md [hunk 2]
```diff
@@ -98,3 +98,4 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Added
 
 -   Added `createAccessList` functionality ( #5780 )
+-   Added support of `safe` and `finalized` block tags (#5823)
```
Comment (Spacesai1or) [inline packages/web3-eth/CHANGELOG.md:101]: For `web3-eth` and `web3-rpc-methods` there wasn't any code changes to add support for the new tags, they just inherited support from other packages that were updated

Comment (Muhammad-Altabba) [reply packages/web3-eth/CHANGELOG.md:101]: @spacesailor24 
I think the whole idea of the CHANGELOG.md is to inform the user about any change in the usage experience. And for the user, it does not make any difference if these changes came from updating some code in the package, or if the changes came from updating a dependency. So, if there was any user-experience change that is introduced in a specific version, then those must be mentioned in the CHANGELOG.md. 

In short, from the user perspective, before this update, it was not possible to pass `safe` or `finalized` block tags. And this change in behavior is introduced in this specific version. And this has to be documented here.

File: packages/web3-utils/src/validation.ts [hunk 3]
```diff
@@ -109,46 +110,37 @@ export const isTopicInBloom = isTopicInBloomValidator;
...
-	if (
-		typeof blockA === 'string' &&
-		!(
-			blockA === 'genesis' ||
-			blockA === 'earliest' ||
-			blockA === 'pending' ||
-			blockA === 'latest'
-		)
-	)
+	if (typeof blockA === 'string' && !(isBlockTag(blockA) || blockA === 'genesis'))
```
Comment (Muhammad-Altabba) [inline packages/web3-utils/src/validation.ts]: I think we need to remove `genesis` as in https://github.com/web3/web3.js/pull/3594/files

File: packages/web3-utils/src/validation.ts [hunk 4]
```diff
@@ -109,46 +110,36 @@ export const isTopicInBloom = isTopicInBloomValidator;
  */
 export const compareBlockNumbers = (blockA: Numbers, blockB: Numbers) => {
```
Comment (Muhammad-Altabba) [inline packages/web3-utils/src/validation.ts]: What about using the type `BlockNumberOrTag` here?
```suggestion
export const compareBlockNumbers = (BlockNumberOrTag: Numbers, BlockNumberOrTag: Numbers) => {
```

File: packages/web3-types/src/eth_types.ts [hunk 5]
```diff
@@ -45,8 +45,11 @@ export enum BlockTags {
 	EARLIEST = 'earliest',
 	LATEST = 'latest',
 	PENDING = 'pending',
+	SAFE = 'safe',
+	FINALIZED = 'finalized',
 }
-export type BlockTag = 'earliest' | 'latest' | 'pending';
+export type BlockTag = `${BlockTags}`;
+
 export type BlockNumberOrTag = Numbers | BlockTag;
```
Comment (Muhammad-Altabba) [inline packages/web3-types/src/eth_types.ts:53]: Since `Numbers` could be any string. I suggest we use this:
```suggestion
export type BlockNumberOrTag = number | bigint | BlockTag;
```

File: packages/web3-utils/src/validation.ts [hunk 6]
```diff
@@ -109,46 +110,36 @@ export const isTopicInBloom = isTopicInBloomValidator;
...
-	if (blockA === 'latest') {
+	}
+	if (blockA === 'earliest') {
+		// b !== a, thus a < b
+		return -1;
+	}
+	if (blockB === 'earliest') {
+		// b !== a, thus a > b
+		return 1;
+	}
```
Comment (Muhammad-Altabba) [inline packages/web3-utils/src/validation.ts]:
```suggestion
	if (blockA === 'earliest' || blockA === 0) {
		// b !== a, thus a < b
		return -1;
	}
	if (blockB === 'earliest' || blockB === 0) {
		// b !== a, thus a > b
		return 1;
	}
```

File: packages/web3-utils/src/validation.ts [hunk 7]
```diff
@@ -162,8 +153,22 @@ export const compareBlockNumbers = (blockA: Numbers, blockB: Numbers) => {
 	if (blockB === 'pending') {
 		return -1;
 	}
+	if (blockA === 'safe' && blockB === 0) {
+		return 1;
+	}
+	if (blockA === 0 && blockB === 'safe') {
+		return -1;
+	}
```
Comment (Muhammad-Altabba) [inline packages/web3-utils/src/validation.ts]: I think, those could be removed if we apply the suggestion above at lines 127 and 131.
```suggestion
```

File: packages/web3-utils/src/validation.ts [hunk 8]
```diff
@@ -109,46 +110,36 @@ export const isTopicInBloom = isTopicInBloomValidator;
...
-			blockB === 'latest'
-		)
-	)
-		throw new InvalidBlockError(blockB);
+	if (blockA === 'genesis' || blockB === 'genesis')
+		throw new InvalidBlockError('Genesis tag not supported'); // for more specific error message
+	if (typeof blockA === 'string' && !isBlockTag(blockA)) throw new InvalidBlockError(blockA);
+	if (typeof blockB === 'string' && !isBlockTag(blockB)) throw new InvalidBlockError(blockB);
+
+	// Increasing order:  earliest, safe, (finalized ~ latest), pending
```
Comment (Muhammad-Altabba) [inline packages/web3-utils/src/validation.ts]: It seems this is the order according to https://www.alchemy.com/overviews/ethereum-commitment-levels and https://hackmd.io/@prysmaticlabs/finality and so, the conditions below need to be updated (I think the conditions for `finalized`  needs to be modified.)
```suggestion
	// Increasing order:  earliest, finalized, safe (justified), latest, pending
```

File: packages/web3-utils/src/validation.ts [hunk 9]
```diff
@@ -109,46 +110,36 @@ export const isTopicInBloom = isTopicInBloomValidator;
...
+	// safe vs block-num cant be compared as block number provided can be on left or right side of safe tag, until safe tag block number is extracted and compared
```
Comment (Muhammad-Altabba) [inline packages/web3-utils/src/validation.ts:119]:
```suggestion
	// Note: Any block tag vs block-num cant be compared as the block number provided can be on the left or the right side of that block tag, until that tag's block number is extracted and compared. 
	//   For example, `pending` is larger than all the produced blocks. But it is smaller than any number of a block that is not yet produced. ex. `pending` is smaller than `Number.MAX_VALUE`.
```

File: packages/web3-utils/src/validation.ts [hunk 10]
```diff
@@ -162,8 +153,22 @@ export const compareBlockNumbers = (blockA: Numbers, blockB: Numbers) => {
...
+	}
+	if (blockA === 0 && blockB === 'safe') {
+		return -1;
+	}
+	if (blockA === 'safe' || blockB === 'safe') {
+		// either a or b is "safe" and the other one did not fall into any of the conditions above, so the other one is a number
+		throw new InvalidBlockError(
+			`Cannot compare safe tag with ${blockA === 'safe' ? blockB : blockA}`,
+		);
+	}
```
Comment (Muhammad-Altabba) [inline packages/web3-utils/src/validation.ts]: I think, any block-tag vs block-num cant be compared as the block number provided can be on the left or the right side of that block tag, until that tag's block number is extracted and compared. 
For example, `pending` is larger than all the produced blocks. But it is smaller than any number of a block that is not yet produced. ex. `pending` is smaller than `Number.MAX_VALUE`.

```suggestion
	if (isBlockTag(blockA) || isBlockTag(blockB)) {
		// either a or b is a blocktag but the other one did not fall into any of the conditions above, so the other one is a number
		throw new InvalidBlockError(
			`Cannot compare a block tag with a number. blockA: ${blockA}, blockB: ${blockB}`,
		);
	}
```

**Review by Muhammad-Altabba — COMMENTED** [review]
I have a few points, please.

File: packages/web3-types/src/eth_types.ts (hunk 5 above)
Comment (jdevcs) [reply packages/web3-types/src/eth_types.ts:53]: Numbers already includes bigint https://github.com/web3/web3.js/blob/f3b8ca88eb34dbf73f8e3dc9f5995c4c1a54fa87/packages/web3-types/src/primitives_types.ts#L20 so this change is not required.

**Review by avkos — APPROVED** [review]

File: packages/web3-utils/src/validation.ts (hunk 4 above)
Comment (jdevcs) [reply packages/web3-utils/src/validation.ts]: you mean: `compareBlockNumbers = (blockA: BlockNumberOrTag, blockB: BlockNumberOrTag)` and not `(BlockNumberOrTag: Numbers, BlockNumberOrTag: Numbers) ` , done

File: packages/web3-types/src/eth_types.ts (hunk 5 above)
Comment (jdevcs) [reply packages/web3-types/src/eth_types.ts:53]: Since BlockNumberOrTag is used in many other places of lib and uses can pass hex string to other functions so this shouldnt be changed.

File: packages/web3-utils/src/validation.ts (hunk 6 above)
Comment (jdevcs) [reply packages/web3-utils/src/validation.ts]: this was already catered in cases below, in case A or B is 0 and other was valid block tag,. but good to check earlier in logic.

File: packages/web3-utils/src/validation.ts (hunk 8 above)
Comment (jdevcs) [reply packages/web3-utils/src/validation.ts]: modified

File: packages/web3-utils/src/validation.ts (hunk 9 above)
Comment (jdevcs) [reply packages/web3-utils/src/validation.ts:119]: This change is not required, as in some case block tag vs number cases are supported. like earliest vs any number etc

File: packages/web3-utils/src/validation.ts (hunk 10 above)
Comment (jdevcs) [reply packages/web3-utils/src/validation.ts]: > I think, any block-tag vs block-num cant be compared as the block number provided can be on the left or the right side of that block tag, until that tag's block number is extracted and compared.

in some cases we can compare tag with number as:
earliest vs num
pending vs num, ( as pending means it will always come after mined block )

Comment (jdevcs) [reply packages/web3-utils/src/validation.ts]: https://github.com/web3/web3.js/issues/5836
